include/deal.II/fe/fe_interface_values.h

```diff
@@ -0,0 +1,686 @@
+   */
+  const std::vector<Tensor<1, spacedim>> &
+  get_normal_vectors() const;
+
+  /**
+   * Return a reference to the quadrature object in use.
+   */
+  const Quadrature<dim - 1> &
+  get_quadrature() const;
+
+  /**
+   * Return a reference to the quadrature points in real space.
+   *
+   * @note This call requires a prior call to reinit() and having
+   * the @p update_quadrature_points flag set in the constructor.
+   */
+  const std::vector<Point<spacedim>> &
+  get_quadrature_points() const;
+
+  /**
+   * Return the update flags set.
+   */
+  const UpdateFlags
+  get_update_flags() const;
+
+  /**
+   * Return the number of DoFs (or shape functions) on the current interface.
+   *
+   * @note This number is only available after a call to reinit() and can change
+   * from one call to reinit() to the next.
+   */
+  unsigned
+  n_current_interface_dofs() const;
+
+  /**
+   * Return if the current interface is a boundary face or an internal
+   * face with two adjacent cells.
+   *
+   * See the corresponding reinit() functions for details.
+   */
+  bool
+  at_boundary() const;
+
+  /**
+   * Return the set of joint DoF indices. This includes indices from both cells.
+   */
+  std::vector<types::global_dof_index>
+  get_interface_dof_indices() const;
+
+  /**
+   * Convert an interface dof index into the corresponding local DoF indices of
+   * the two cells. If an interface DoF is only active on one of the
+   * cells, the other index will be numbers::invalid_unsigned_int.
+   *
+   * For discontinuous finite elements each interface dof will correspond to
+   * exactly one DoF index.
+   */
+  std::array<unsigned int, 2>
+  interface_dof_to_dof_indices(const unsigned int interface_dof_index) const;
+
+  /**
+   * Return the normal in a given quadrature point.
+   *
+   * The normal points in outwards direction as seen from the first cell of
+   * this interface.
+   */
+  Tensor<1, spacedim>
+  normal(const unsigned int q_point_index) const;
+
+  /**
+   * Return component @p component of the value of the shape function
+   * with interface dof index @p interface_dof_index in
+   * quadrature point @p q_point.
+   *
+   * The argument @p use_upstream_value selects between the upstream value and
+   * the downstream value as defined by the direction of the normal vector
+   * in this quadrature point. If @p use_upstream_value is true, the shape
+   * functions from the first cell of the interface is used.
+   *
+   * In other words, this function returns the limit of the value of the shape
+   * function in the given quadrature point when approaching it from one of the
+   * two cells of the interface.
+   *
+   * @note This function is typically used to pick the upstream or downstream
+   * value based on a direction. This can be achieved by using
+   * <code>(direction * normal)>0</code> as the first argument of this
+   * function.
+   */
+  double
+  shape_value(const bool         use_upstream_value,
```

Comment:
I think the function would be easier to understand if you named the argument something like `here_or_there`. The term "upstream" implies a direction, but no direction is provided -- you assume that the "stream direction" is the normal vector from cell 0, but then the (good) example in the last paragraph introduces a different direction vector.